File: server/app/download.py
"""Track-download endpoint for loupe — yt-dlp, optional at runtime.

Implements the HTTP contract the web `http-track-source` adapter speaks:

    POST /download             body = {"url": "..."} (application/json)
      -> 200 application/x-ndjson, streamed line by line:
           {"type":"progress","phase":"downloading"|"transcoding","fraction":0..1}
           {"type":"done","ref","title","duration"?,"uploader"?}   (last line)
           {"type":"error","message"}                              (on failure)

The heavy lifting is **yt-dlp** (YouTube / SoundCloud extraction); this module is
only the glue + transport. `main` imports it lazily so a host without yt-dlp
still serves the rest — `/download` then answers with an NDJSON error line.

The fetched audio is parked in the SAME content-addressed `/audio` store the
project blobs use (via `projects.store_audio`), so the web client resolves the
bytes with a plain `GET /audio/{ref}`. We ask yt-dlp for `bestaudio[ext=m4a]`
so the common case needs no ffmpeg re-encode — the downloaded m4a/AAC decodes
straight in the browser's `decodeAudioData`.

Single-user, localhost — no auth. The host allowlist mirrors the core's
`isSupportedSourceUrl`: yt-dlp is a powerful fetcher, so we never point it at an
arbitrary URL.
"""

from __future__ import annotations

import json
import logging
import queue
import tempfile
import threading
import time
from collections.abc import Iterator
from pathlib import Path
from urllib.parse import urlparse

import yt_dlp
from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import StreamingResponse

from .limits import (
    MAX_MANIFEST_BYTES,
    MAX_UPLOAD_BYTES,
    concurrency_slots,
    read_capped_json,
    seconds_env,
)
from .projects import store_audio

logger = logging.getLogger("loupe.download")
router = APIRouter()

# yt-dlp holds a network pipe and (on the ffmpeg fallback) a CPU; N parallel
# /download calls would also fill the tmp dir N times over. One at a time by
# default; env-tunable. Same worker-side acquisition as /separate.
_download_semaphore = threading.BoundedSemaphore(
    concurrency_slots("LOUPE_MAX_CONCURRENT_DOWNLOADS")
)

# A wedged yt-dlp (network black hole, stuck extractor) must not suspend the
# stream's threadpool thread forever — cap the wait between two worker events.
DOWNLOAD_TIMEOUT_SECONDS = seconds_env("LOUPE_DOWNLOAD_TIMEOUT_SECONDS", 900)

# Mirror of the core's supported-source policy: only these hosts (and their
# subdomains) are fetchable. Keeps yt-dlp from being used as an open proxy.
# Kept in sync with packages/core/src/application/supported-source.ts and the
# desktop shell's Rust copy (packages/desktop/src-tauri/src/download.rs).
_SUPPORTED_HOSTS = ("youtube.com", "youtu.be", "soundcloud.com")


def _is_supported(url: str) -> bool:
    try:
        parsed = urlparse(url)
    except ValueError:
        return False
    if parsed.scheme not in ("http", "https") or not parsed.hostname:
        return False
    host = parsed.hostname.lower()
    return any(host == h or host.endswith(f".{h}") for h in _SUPPORTED_HOSTS)


def _event(payload: dict) -> bytes:
    return (json.dumps(payload) + "\n").encode("utf-8")


def progress_fraction(status: dict) -> float | None:
    """The download completion in [0, 1] from a yt-dlp progress hook.

    None unless we're downloading and have both a total and a done count — a
    missing/zero total (yt-dlp not yet knowing the size) yields no fraction.
    """
    if status.get("status") != "downloading":
        return None
    total = status.get("total_bytes") or status.get("total_bytes_estimate")
    done = status.get("downloaded_bytes")
    if total and done is not None:
        return min(1.0, done / total)
    return None


def _make_options(out_dir: Path, on_progress) -> dict:
    """yt-dlp options: audio-only, no playlist, m4a-preferred, progress-hooked."""
    return {
        "format": "bestaudio[ext=m4a]/bestaudio",
        "outtmpl": str(out_dir / "%(id)s.%(ext)s"),
        "noplaylist": True,
        "quiet": True,
        "no_warnings": True,
        "progress_hooks": [on_progress],
        # The audio store would refuse a blob over the upload cap anyway —
        # refuse it before it fills the tmp dir, which sits outside the quota.
        "max_filesize": MAX_UPLOAD_BYTES,
        # A dead pipe must raise inside yt-dlp: the stream deadline below frees
        # our threadpool thread, but only the worker can free its own slot.
        "socket_timeout": 30,
    }


def _extract(url: str, out_dir: Path, on_progress) -> dict:
    """Download the track once, returning yt-dlp's info dict."""
    # yt-dlp types its options as a private TypedDict and returns a private
    # `_InfoDict`; we treat both as plain dicts at the boundary.
    with yt_dlp.YoutubeDL(_make_options(out_dir, on_progress)) as ydl:  # pyright: ignore[reportArgumentType]
        return ydl.extract_info(url, download=True)  # pyright: ignore[reportReturnType]


def _download(url: str, out_dir: Path, events: queue.Queue) -> None:
    """Fetch on a worker thread, pushing progress onto `events`.

    yt-dlp's progress hook fires on the same thread; we translate it into
    `downloading` fractions. A `('done', info)` or `('error', message)` tuple
    ends the stream. A stale extractor (a service changed its player) surfaces a
    clear message telling the operator to run `pip install -U yt-dlp` themselves —
    a request never triggers a package install (that would be remote code pulled
    over the network and executed in-process).
    """

    def on_progress(status: dict) -> None:
        fraction = progress_fraction(status)
        if fraction is not None:
            events.put(("progress", fraction))

    try:
        with _download_semaphore:
            info = _extract(url, out_dir, on_progress)
        events.put(("done", info))
    except yt_dlp.utils.DownloadError:  # pyright: ignore[reportAttributeAccessIssue]
        events.put(
            (
                "error",
                "download failed — the extractor may be out of date; "
                "run `pip install -U yt-dlp` in the server venv and retry",
            )
        )
    except Exception:  # noqa: BLE001  # logged server-side, generic to the client
        logger.exception("download failed")
        events.put(("error", "download failed"))


def _download_stream(url: str) -> Iterator[bytes]:
    """Run the download, streaming live NDJSON progress then the parked-ref done.

    FastAPI iterates this sync generator in a threadpool; the fetch runs on a
    further worker thread so we can yield its progress as it arrives (mirrors
    `/separate`).
    """
    if not _is_supported(url):
        yield _event({"type": "error", "message": f"unsupported source URL: {url}"})
        return

    yield _event({"type": "progress", "phase": "downloading", "fraction": 0.0})

    with tempfile.TemporaryDirectory(prefix="loupe-download-") as tmp:
        out_dir = Path(tmp)
        events: queue.Queue = queue.Queue()
        worker = threading.Thread(target=_download, args=(url, out_dir, events), daemon=True)
        worker.start()

        # A single wall-clock budget for the whole fetch: a per-event timeout
        # would be reset forever by a trickling download's progress stream.
        deadline = time.monotonic() + DOWNLOAD_TIMEOUT_SECONDS
        info = None
        while True:
            remaining = deadline - time.monotonic()
            try:
                if remaining <= 0:
                    raise queue.Empty
                kind, payload = events.get(timeout=remaining)
            except queue.Empty:
                yield _event({"type": "error", "message": "download timed out"})
                return
            if kind == "progress":
                yield _event({"type": "progress", "phase": "downloading", "fraction": payload})
            elif kind == "error":
                yield _event({"type": "error", "message": payload})
                return
            else:  # done
                info = payload
                break

        # Extraction/parking step — signalled as `transcoding` so the client's
        # progress bar completes even though the common m4a path needs no
        # re-encode.
        yield _event({"type": "progress", "phase": "transcoding", "fraction": 1.0})

        files = list(out_dir.iterdir())
        if not files:
            # yt-dlp honours `max_filesize` by silently skipping the download,
            # so an empty dir usually means the track blew the size cap.
            yield _event(
                {
                    "type": "error",
                    "message": "download produced no file — the track may exceed the size cap",
                }
            )
            return
        try:
            ref = store_audio(files[0].read_bytes())
        except HTTPException as exc:
            # The HTTP 200 is already committed (streaming) — a refusal such as
            # the store quota must arrive as an NDJSON error, not a dead stream.
            yield _event({"type": "error", "message": str(exc.detail)})
            return

    done = {"type": "done", "ref": ref, "title": info.get("title") or "Sans titre"}
    if info.get("duration") is not None:
        done["duration"] = float(info["duration"])
    if info.get("uploader"):
        done["uploader"] = info["uploader"]
    yield _event(done)


@router.post("/download")
async def download(request: Request) -> StreamingResponse:
    _, body = await read_capped_json(request, MAX_MANIFEST_BYTES)
    url = body.get("url", "") if isinstance(body, dict) else ""
    return StreamingResponse(
        _download_stream(url),
        media_type="application/x-ndjson",
    )

File: crates/loupe-server/src/download.rs
//! `POST /download` — the NDJSON contract the web `http-track-source`
//! adapter speaks, backed by the shared `loupe-download` engine (D4.a):
//!
//! ```text
//! POST /download        body = {"url": "..."} (application/json)
//!   -> 200 application/x-ndjson, streamed line by line:
//!        {"type":"progress","phase":"downloading"|"transcoding","fraction":0..1}
//!        {"type":"done","ref","title","duration"?,"uploader"?}   (last line)
//!        {"type":"error","code","message"}                       (on failure)
//!
//! The error `code` (`unsupported` | `timeout` | `extractor-stale` |
//! `store-quota` | `unknown`) is the UI contract (AV.1): the client maps it to
//! translated copy and keeps `message` for the console.
//! ```
//!
//! The engine is a seam (`DownloadEngine`) so tests drive the route with a
//! fake; the binary wires `loupe_download::download_track`. The fetched audio
//! is parked in the SAME content-addressed `/audio` store the project blobs
//! use, so the web client resolves the bytes with a plain `GET /audio/{ref}`
//! — never a forked second store.

use crate::audio_store::{StoreError, QUOTA_MESSAGE};
use crate::AppState;
use axum::body::{Body, Bytes};
use axum::extract::State;
use axum::http::{header, StatusCode};
use axum::response::{IntoResponse, Response};
use loupe_download::{DownloadError, DownloadedTrack, ProgressFn};
use std::future::Future;
use std::path::PathBuf;
use std::pin::Pin;
use std::sync::Arc;
use tokio::sync::mpsc::UnboundedSender;
use tokio::sync::Notify;
use tokio_stream::wrappers::UnboundedReceiverStream;
use tokio_stream::StreamExt;

pub type TrackFuture = Pin<Box<dyn Future<Output = Result<DownloadedTrack, DownloadError>> + Send>>;

/// The download seam: same signature as `loupe_download::download_track`,
/// object-safe so the route can be exercised with a fake engine.
pub trait DownloadEngine: Send + Sync {
  fn download(
    &self,
    data_dir: PathBuf,
    url: String,
    on_progress: Arc<ProgressFn>,
    cancel: Arc<Notify>,
  ) -> TrackFuture;
}

/// The real engine — the one shared implementation of D4.a.
pub struct YtDlpEngine;

impl DownloadEngine for YtDlpEngine {
  fn download(
    &self,
    data_dir: PathBuf,
    url: String,
    on_progress: Arc<ProgressFn>,
    cancel: Arc<Notify>,
  ) -> TrackFuture {
    Box::pin(
      async move { loupe_download::download_track(&data_dir, &url, on_progress, cancel).await },
    )
  }
}

/// Notifies the engine's cancel handle when the response stream is dropped —
/// a client that disconnects mid-download kills its own yt-dlp child instead
/// of leaving it running to completion for nobody.
struct CancelOnDrop(Arc<Notify>);

impl Drop for CancelOnDrop {
  fn drop(&mut self) {
    self.0.notify_waiters();
  }
}

fn line(payload: serde_json::Value) -> String {
  format!("{payload}\n")
}

/// The NDJSON error line: the `code` is what the client's copy switches on,
/// the raw English `message` is console-only material.
fn error_line(code: &str, message: &str) -> serde_json::Value {
  serde_json::json!({"type": "error", "code": code, "message": message})
}

fn send(events: &UnboundedSender<String>, payload: serde_json::Value) {
  // A closed channel just means the client is gone — nothing to report to.
  let _ = events.send(line(payload));
}

pub async fn download(State(state): State<Arc<AppState>>, body: Bytes) -> Response {
  let Ok(parsed) = serde_json::from_slice::<serde_json::Value>(&body) else {
    return (StatusCode::BAD_REQUEST, "body is not JSON").into_response();
  };
  let url = parsed
    .get("url")
    .and_then(|value| value.as_str())
    .unwrap_or("")
    .to_owned();

  let (events, receiver) = tokio::sync::mpsc::unbounded_channel::<String>();
  let cancel = Arc::new(Notify::new());
  tokio::spawn(run_download(state, url, events, cancel.clone()));

  // The HTTP 200 is committed up front (streaming): every later refusal —
  // engine failure, store quota, timeout — arrives as an NDJSON error line,
  // never a dead stream or a late status change.
  let guard = CancelOnDrop(cancel);
  let stream = UnboundedReceiverStream::new(receiver).map(move |chunk| {
    let _keep_alive = &guard;
    Ok::<_, std::convert::Infallible>(Bytes::from(chunk))
  });
  (
    [(header::CONTENT_TYPE, "application/x-ndjson")],
    Body::from_stream(stream),
  )
    .into_response()
}

async fn run_download(
  state: Arc<AppState>,
  url: String,
  events: UnboundedSender<String>,
  cancel: Arc<Notify>,
) {
  let progress_events = events.clone();
  let on_progress: Arc<ProgressFn> = Arc::new(move |phase, fraction| {
    let _ = progress_events.send(line(
      serde_json::json!({"type": "progress", "phase": phase, "fraction": fraction}),
    ));
  });

  // One wall-clock budget for the whole run, queue wait included — parity
  // with the Python stream's single deadline (a trickle does not re-arm it).
  let work = async {
    // yt-dlp holds a network pipe (and on the ffmpeg fallback a CPU): N
    // parallel downloads would also fill the temp dir N times over. One at a
    // time by default, env-tunable — same policy as the Python semaphore.
    let Ok(_permit) = state.download_slots.acquire().await else {
      return None;
    };
    if events.is_closed() {
      return None; // the client left while we were queued — skip the fetch
    }
    Some(
      state
        .engine
        .download(state.config.data_dir.clone(), url, on_progress, cancel)
        .await,
    )
  };
  match tokio::time::timeout(state.config.download_timeout, work).await {
    Err(_) => send(&events, error_line("timeout", "download timed out")),
    Ok(None) => {}
    Ok(Some(Err(error))) => send(&events, error_line(error.code.as_str(), &error.message)),
    Ok(Some(Ok(track))) => match park(&state, &track).await {
      Ok(done) => send(&events, done),
      Err((code, message)) => send(&events, error_line(code, &message)),
    },
  }
}

/// Move the fetched file into the content-addressed store and build the
/// `done` event; the per-download temp dir goes away either way. A failure is
/// `(code, message)` — the store quota is the one case the user can act on.
async fn park(
  state: &Arc<AppState>,
  track: &DownloadedTrack,
) -> Result<serde_json::Value, (&'static str, String)> {
  let audio_path = state.config.data_dir.join(&track.relative_path);
  let temp_dir = audio_path.parent().map(PathBuf::from);
  let store = state.store.clone();
  let stored = tokio::task::spawn_blocking(move || {
    let bytes = std::fs::read(&audio_path)
      .map_err(|e| ("unknown", format!("downloaded file unreadable: {e}")))?;
    store.store(&bytes).map_err(|error| match error {
      StoreError::QuotaExceeded => ("store-quota", QUOTA_MESSAGE.to_owned()),
      StoreError::Io(message) => ("unknown", format!("audio store error: {message}")),
    })
  })
  .await
  .map_err(|e| ("unknown", format!("audio store task failed: {e}")));
  if let Some(dir) = temp_dir {
    let _ = tokio::fs::remove_dir_all(dir).await;
  }
  let reference = stored??;

  let mut done = serde_json::json!({
    "type": "done",
    "ref": reference,
    "title": track.title,
  });
  if let Some(duration) = track.duration_seconds {
    done["duration"] = duration.into();
  }
  if let Some(uploader) = track.uploader.as_deref().filter(|u| !u.is_empty()) {
    done["uploader"] = uploader.into();
  }
  Ok(done)
}
